File: core/r/tests/testthat/test-proportions.R
# The proportions recipe: chi-squared test on a 2x2 table of arm x response.
test_that("recipe_proportions runs a chi-squared test on a 2x2 table", {
  df <- data.frame(
    arm = c(rep("control", 20), rep("treated", 20)),
    response = c(rep("no", 15), rep("yes", 5), rep("no", 5), rep("yes", 15)),
    stringsAsFactors = FALSE)
  opt <- list(recipe = "proportions", data = "p.csv", x = "arm", y = "response", test = "auto")
  spec <- spec_from_opt(opt)

  res <- recipe_proportions(df, spec)

  expect_equal(res$test_meta$name, "Pearson's chi-squared test")
  expect_true(res$test_meta$p_value < 0.01)
  expect_equal(nrow(res$stats), 4)
})

File: core/python/tests/test_proportions.py
from types import SimpleNamespace

import pandas as pd

from pubplot.recipes.proportions import recipe_proportions
from pubplot.spec import spec_from_args


def test_proportions_chisq_2x2():
    df = pd.DataFrame({
        "arm": ["control"] * 20 + ["treated"] * 20,
        "response": (["no"] * 15 + ["yes"] * 5) + (["no"] * 5 + ["yes"] * 15),
    })
    a = SimpleNamespace(recipe="proportions", data="p.csv", x="arm", y="response", test="auto")
    spec = spec_from_args(a)

    res = recipe_proportions(df, spec)

    assert res["test_meta"]["name"] == "Pearson's chi-squared test"
    assert res["test_meta"]["p_value"] < 0.01
    assert res["test_meta"]["odds_ratio"] is not None
    assert len(res["stats"]) == 4  # 2 groups x 2 outcomes
